src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -610,23 +955,86 @@ internal class PartialCaseDecision
             public readonly SyntaxNode Syntax;
             public readonly ImmutableArray<BoundDagDecision> Decisions;
             public readonly ImmutableArray<(BoundExpression, BoundDagTemp)> Bindings;
-            public readonly BoundExpression WhereClause;
+            public readonly BoundExpression WhenClause;
             public readonly LabelSymbol CaseLabel;
             public PartialCaseDecision(
                 int Index,
                 SyntaxNode Syntax,
                 ImmutableArray<BoundDagDecision> Decisions,
                 ImmutableArray<(BoundExpression, BoundDagTemp)> Bindings,
-                BoundExpression WhereClause,
+                BoundExpression WhenClause,
                 LabelSymbol CaseLabel)
             {
                 this.Index = Index;
                 this.Syntax = Syntax;
                 this.Decisions = Decisions;
                 this.Bindings = Bindings;
-                this.WhereClause = WhereClause;
+                this.WhenClause = WhenClause;
                 this.CaseLabel = CaseLabel;
             }
+
+            public override bool Equals(object obj)
+            {
+                return Equals(obj as PartialCaseDecision);
+            }
+
+            public bool Equals(PartialCaseDecision other)
+            {
+                // We do not include Syntax, Bindings, WhereClause, or CaseLabel
+                // because once the Index is the same, those must be the same too.
+                if (other == null || this.Index != other.Index || this.Decisions.Length != other.Decisions.Length)
+                {
+                    return false;
+                }
+
+                for (int i = 0; i < Decisions.Length; i++)
+                {
+                    if (!SameDecision(Decisions[i], other.Decisions[i]))
+                    {
+                        return false;
+                    }
+                }
+
+                return true;
```

Comment:
`return this.Decisions.SequenceEqual(other.Decisions, (a, b) => SameDecision(a, b));`